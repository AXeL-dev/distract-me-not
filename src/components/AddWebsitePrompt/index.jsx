import React, { Component } from 'react';
import { Pane, toaster } from 'evergreen-ui';
import { translate } from 'helpers/i18n';
import { TextField } from 'components';
import { Mode, blockUrl } from 'helpers/block';
import { isUrl } from 'helpers/url';
import queryString from 'query-string';

export class AddWebsitePrompt extends Component {
  constructor(props) {
    super(props);
    const params = props.location ? queryString.parse(props.location.search) : {};
    this.mode = params.mode || Mode.blacklist;
    this.tabId = +params.tabId || null;
    this.state = {
      url: params.url || '',
      disabled: false,
    };
  }

  componentDidMount() {
    document.addEventListener('keyup', this.handleKeyUp);
  }

  componentWillUnmount() {
    document.removeEventListener('keyup', this.handleKeyUp);
  }

  handleKeyUp = (event) => {
    if (event.code === 'Escape') {
      window.close();
    }
  };

  handleChange = (event) => {
    this.setState({ url: event.target.value });
  };

  handleSubmit = async () => {
    if (!isUrl(this.state.url)) {
      toaster.danger(translate('urlIsNotValid'), {
        id: 'add-website-toaster',
      });
    } else {
      this.setState({ disabled: true });
      try {
        const blocked = await blockUrl(this.state.url, this.mode, this.tabId);
        if (blocked) {
          window.close();
        } else {
          throw new Error(translate('urlAlreadyExists'));
        }
      } catch (error) {
        toaster.danger(error.message, { id: 'add-website-toaster' });
        this.setState({ disabled: false });
      }
    }
  };

  render() {
    return (
      <Pane
        display="flex"
        flex={1}
        width="100%"
        height="100%"
        minWidth={500}
        minHeight={120}
        padding={30}
        alignItems="center"
        justifyContent="center"
      >
        <TextField
          placeholder={translate('urlExample')}
          hint={translate('addWebsiteHint')}
          value={this.state.url}
          hasButton={true}
          buttonLabel={translate('add')}
          onSubmit={this.handleSubmit}
          onChange={this.handleChange}
          disabled={this.state.disabled}
          required
        />
      </Pane>
    );
  }
}
