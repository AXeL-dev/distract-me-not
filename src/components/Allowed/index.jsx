import React, { Component } from 'react';
import { filter } from 'fuzzaldrin-plus';
import { format } from 'date-fns';
import {
  Pane,
  Table,
  Popover,
  Position,
  Menu,
  Text,
  IconButton,
  MoreIcon,
  ArrowUpIcon,
  ArrowDownIcon,
  CaretDownIcon,
  TextDropdownButton,
  TickCircleIcon,
  RefreshIcon,
  TimeIcon,
} from 'evergreen-ui';
import { sendMessage } from 'helpers/webext';
import { translate } from 'helpers/i18n';
import { getHostname } from 'helpers/url';
import { isDevEnv } from 'helpers/debug';
import { now } from 'helpers/date';
import './styles.scss';

const Order = {
  NONE: 'NONE',
  ASC: 'ASC',
  DESC: 'DESC',
};

const allowedHostsDataset = [
  {
    time: 60000 * 10,
    startedAt: new Date().getTime(),
    hostname: getHostname('https://www.website1.com'),
  },
  {
    time: 60000 * 15,
    startedAt: new Date().getTime(),
    hostname: getHostname('https://www.website2.com'),
  },
  {
    time: 60000 * 30,
    startedAt: new Date().getTime(),
    hostname: getHostname('https://www.website3.com'),
  },
];

export class Allowed extends Component {
  constructor(props) {
    super(props);
    this.state = {
      list: [],
      searchQuery: '',
      orderedColumn: 1,
      scrollToIndex: null,
      showRemainingTime: true,
      currentTime: now(true),
      ordering: Order.NONE,
    };
  }

  componentDidMount() {
    this.fetchAllowedHosts();
    this.interval = setInterval(() => this.setState({ currentTime: now(true) }), 1000);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  fetchAllowedHosts = (scrollToTop = false) => {
    sendMessage('getTmpAllowed').then((allowed) => {
      const data = isDevEnv ? allowedHostsDataset : allowed;
      this.setState({
        list: this.getOrderedList(data).filter((row) => this.getRemainingTime(row) > 0),
        scrollToIndex: scrollToTop
          ? data.length > 0
            ? 0
            : null
          : this.state.scrollToIndex,
      });
    });
  };

  getRemainingTime = (row) => {
    const endTime = row.startedAt + row.time;
    return endTime > this.state.currentTime ? endTime - this.state.currentTime : 0;
  };

  getOrderedList = (list) => {
    return list ? list.map((item, index) => ({ id: index + 1, ...item })) : [];
  };

  sort = (items) => {
    const { ordering } = this.state;
    // Return if there's no ordering.
    if (ordering === Order.NONE) return items;

    // Get the property to sort each item on.
    // By default use the `hostname` property.
    let propKey = 'hostname';

    return items.sort((a, b) => {
      let aValue = a[propKey];
      let bValue = b[propKey];

      // Support string comparison
      const sortTable = { true: 1, false: -1 };

      // Order ascending (Order.ASC)
      if (this.state.ordering === Order.ASC) {
        return aValue === bValue ? 0 : sortTable[aValue > bValue];
      }

      // Order descending (Order.DESC)
      return bValue === aValue ? 0 : sortTable[bValue > aValue];
    });
  };

  // Filter the items based on the name property.
  filter = (items) => {
    const searchQuery = this.state.searchQuery.trim();

    // If the searchQuery is empty, return the items as is.
    if (searchQuery.length === 0) return items;

    return items.filter((item) => {
      // Use the filter from fuzzaldrin-plus to filter by hostname.
      const result = filter([item.hostname], searchQuery);
      return result.length === 1;
    });
  };

  getIconForOrder = (order) => {
    switch (order) {
      case Order.ASC:
        return ArrowUpIcon;
      case Order.DESC:
        return ArrowDownIcon;
      default:
        return CaretDownIcon;
    }
  };

  handleFilterChange = (value) => {
    this.setState({ searchQuery: value });
  };

  renderColumnSortButton = ({ orderedColumn, label }) => {
    return (
      <Popover
        position={Position.BOTTOM_LEFT}
        minWidth={160}
        content={({ close }) => (
          <Menu>
            <Menu.OptionsGroup
              title={translate('order')}
              options={[
                { label: translate('ascending'), value: Order.ASC },
                { label: translate('descending'), value: Order.DESC },
              ]}
              selected={
                this.state.orderedColumn === orderedColumn ? this.state.ordering : null
              }
              onChange={(value) => {
                this.setState({
                  orderedColumn,
                  ordering: value,
                });
                // Close the popover when you select a value.
                close();
              }}
            />
          </Menu>
        )}
      >
        <TextDropdownButton
          icon={
            this.state.orderedColumn === orderedColumn
              ? this.getIconForOrder(this.state.ordering)
              : CaretDownIcon
          }
          data-testid="sort-button"
        >
          {label}
        </TextDropdownButton>
      </Popover>
    );
  };

  renderColumnTimeButton = ({ label } = {}) => {
    return (
      <Popover
        position={Position.BOTTOM_LEFT}
        minWidth={160}
        content={({ close }) => (
          <Menu>
            <Menu.OptionsGroup
              title={translate('remainingTime')}
              options={[
                { label: translate('show'), value: true },
                { label: translate('hide'), value: false },
              ]}
              selected={this.state.showRemainingTime}
              onChange={(value) => {
                this.setState({ showRemainingTime: value });
                // Close the popover when you select a value.
                close();
              }}
            />
          </Menu>
        )}
      >
        <TextDropdownButton icon={TimeIcon} data-testid="date-button">
          {label}
        </TextDropdownButton>
      </Popover>
    );
  };

  renderHeaderMenu = ({ close }) => {
    return (
      <Menu>
        <Menu.Group>
          <Menu.Item
            icon={RefreshIcon}
            onSelect={() => {
              this.fetchAllowedHosts(true);
              close();
            }}
          >
            {translate('refresh')}
          </Menu.Item>
        </Menu.Group>
      </Menu>
    );
  };

  renderRow = ({ row }) => {
    const remainingTime = this.getRemainingTime(row);

    return (
      <Table.Row key={row.id} height={38}>
        <Table.Cell
          flex={this.state.showRemainingTime ? 1 : 'none'}
          display="flex"
          alignItems="center"
          title={translate('allowed')}
        >
          <TickCircleIcon color="#28a745" size={16} />
          <Text marginLeft={8} size={300} fontWeight={500} data-testid="hostname">
            {row.hostname}
          </Text>
        </Table.Cell>
        {this.state.showRemainingTime && (
          <Table.Cell flex="none">
            <Text marginLeft={8} size={300} fontWeight={500} data-testid="remaining-time">
              {format(new Date(remainingTime), 'HH:mm:ss')}
            </Text>
          </Table.Cell>
        )}
      </Table.Row>
    );
  };

  render() {
    const items = this.filter(this.sort(this.state.list));

    return (
      <Pane padding={16} height="100%">
        <Table border height="100%">
          <Table.Head height={32} padding={0}>
            <Table.SearchHeaderCell
              onChange={this.handleFilterChange}
              value={this.state.searchQuery}
              placeholder={translate('filter') + '...'}
            />
            <Table.HeaderCell flex="none">
              {this.renderColumnSortButton({ orderedColumn: 1 })}
            </Table.HeaderCell>
            <Table.HeaderCell flex="none">
              {this.renderColumnTimeButton()}
            </Table.HeaderCell>
            <Table.HeaderCell width={48} flex="none">
              <Popover
                content={({ close }) => this.renderHeaderMenu({ close: close })}
                position={Position.BOTTOM_RIGHT}
                minWidth={160}
              >
                <IconButton
                  icon={MoreIcon}
                  height={24}
                  appearance="minimal"
                  data-testid="list-more-button"
                />
              </Popover>
            </Table.HeaderCell>
          </Table.Head>
          <Table.VirtualBody
            scrollToIndex={items.length > 0 ? this.state.scrollToIndex : undefined}
            height="calc(100% - 32px)"
          >
            {items.map((item) => this.renderRow({ row: item }))}
          </Table.VirtualBody>
        </Table>
      </Pane>
    );
  }
}
