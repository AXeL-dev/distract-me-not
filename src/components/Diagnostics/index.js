export { default as Diagnostics } from './index.jsx';
export { default } from './index.jsx';
