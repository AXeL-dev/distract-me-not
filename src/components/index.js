export * from './Background';
export * from './Blocked';
export * from './Allowed';
export * from './Panel';
export * from './PasswordPrompt';
export * from './AddWebsitePrompt';
export * from './PasteBin';
export * from './Settings';
export * from './Timer';
export * from './Logs';
export * from './Main';
export * from './shared';
