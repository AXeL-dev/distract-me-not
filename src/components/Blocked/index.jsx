import React, { Component, Fragment } from 'react';
import {
  Pane,
  Dialog,
  RadioGroup,
  Button,
  UnlockIcon,
  toaster,
  DuplicateIcon,
} from 'evergreen-ui';
import { translate } from 'helpers/i18n';
import { storage, sendMessage } from 'helpers/webext';
import { debug, isDevEnv, isProdEnv } from 'helpers/debug';
import { isUrl, getValidUrl } from 'helpers/url';
import {
  UnblockOptions,
  isPageReloaded,
  defaultUnblockSettings,
  defaultBlockSettings,
} from 'helpers/block';
import { NumberInput, PasswordPrompt } from 'components';
import queryString from 'query-string';
import copy from 'copy-to-clipboard';
import './styles.scss';

export class Blocked extends Component {
  constructor(props) {
    super(props);
    this.url = props.location ? queryString.parse(props.location.search).url : null;
    if (this.url && isUrl(this.url)) {
      this.url = decodeURIComponent(this.url);
      this.url = getValidUrl(this.url);
    }
    if (isDevEnv && !this.url) {
      this.url = 'https://www.example.com';
    }
    debug.log('url', this.url);
    const defaultUnblockTime = 10; // min
    this.state = {
      message: props.message || translate('defaultBlockingMessage'),
      isBlank: props.isBlank === undefined ? isProdEnv : props.isBlank, // isBlank should be true by default in prod
      hasUnblockButton: props.hasUnblockButton || isDevEnv,
      displayBlockedLink:
        props.displayBlockedLink || isDevEnv || defaultBlockSettings.displayBlockedLink,
      unblockDialog: {
        isShown: false,
        options: this.getUnblockOptions(defaultUnblockTime),
        selected: UnblockOptions.unblockOnce,
        time: defaultUnblockTime,
        requirePassword: props.requirePassword || false,
      },
    };
  }

  getUnblockOptions = (time) => {
    return [
      {
        label: translate('unblockOnce'),
        value: UnblockOptions.unblockOnce,
      },
      {
        label: (
          <Pane display="flex" alignItems="center" gap={10}>
            <span>{translate('unblockFor')}</span>
            <NumberInput
              min={1}
              max={720}
              width={65}
              value={time}
              onChange={(value) =>
                this.updateUnblockDialogState({
                  selected: UnblockOptions.unblockForWhile,
                  time: value,
                })
              }
            />
            <span>{translate('minutes')}</span>
          </Pane>
        ),
        value: UnblockOptions.unblockForWhile,
      },
    ];
  };

  componentDidMount() {
    if (isPageReloaded()) {
      debug.log('page reloaded!');
      sendMessage('isUrlStillBlocked', this.url).then((isUrlStillBlocked) => {
        // Redirect to blocked url if no longer blocked
        if (isUrlStillBlocked === false && this.url) {
          sendMessage('redirectSenderTab', this.url);
          return;
        }
      });
    }
    storage
      .get({
        message: this.state.message,
        displayBlankPage: this.state.isBlank,
        displayBlockedLink: this.state.displayBlockedLink,
        unblock: {
          isEnabled: isDevEnv || defaultUnblockSettings.isEnabled,
          requirePassword: defaultUnblockSettings.requirePassword,
        },
        password: {
          isEnabled: false,
        },
      })
      .then((items) => {
        if (items) {
          this.setState({
            message: items.message.length ? items.message : this.state.message,
            isBlank: items.displayBlankPage,
            displayBlockedLink: items.displayBlockedLink,
            hasUnblockButton: items.unblock.isEnabled,
            unblockDialog: {
              ...this.state.unblockDialog,
              requirePassword:
                items.unblock.isEnabled &&
                items.unblock.requirePassword &&
                items.password.isEnabled,
            },
          });
        }
      });
  }

  updateUnblockDialogState = (state) => {
    this.setState({
      unblockDialog: {
        ...this.state.unblockDialog,
        ...state,
      },
    });
  };

  closeUnblockDialog = () => {
    this.updateUnblockDialogState({ isShown: false });
  };

  openUnblockDialog = () => {
    this.updateUnblockDialogState({ isShown: true });
  };

  unblock = () => {
    this.closeUnblockDialog();
    const params = {
      url: this.url,
      option: this.state.unblockDialog.selected,
      time: this.state.unblockDialog.time,
    };
    debug.log('unblocking:', params);
    if (this.url) {
      //window.location.replace(this.url);
      sendMessage('unblockSenderTab', params);
    }
  };

  copyBlockedLink = () => {
    if (copy(this.url)) {
      toaster.success(translate('copiedToClipboard'), {
        id: 'blocked-toaster',
      });
    }
  };

  render() {
    return (
      <Fragment>
        {!this.state.isBlank && (
          <Fragment>
            <div className="distract-cursor distract-select distract-overlay-container">
              <div className="distract-cursor distract-select distract-overlay">
                <div className="distract-cursor distract-select distract-info-container">
                  <span className="distract-cursor distract-select distract-overlay-top-text">
                    {this.state.message}
                  </span>
                  {this.state.displayBlockedLink && (
                    <span className="distract-blocked-link">
                      <input type="text" value={this.url || ''} readOnly />
                      <button
                        className="copy"
                        title={translate('copy')}
                        onClick={this.copyBlockedLink}
                      >
                        <DuplicateIcon />
                      </button>
                    </span>
                  )}
                  <div className="distract-cursor distract-select distract-overlay-img"></div>
                  {this.state.hasUnblockButton && (
                    <button className="unblock" onClick={this.openUnblockDialog}>
                      {translate('unblock')}
                    </button>
                  )}
                </div>
              </div>
            </div>
            <Dialog
              isShown={this.state.unblockDialog.isShown}
              onCloseComplete={this.closeUnblockDialog}
              shouldCloseOnOverlayClick={false}
              //hasHeader={false}
              hasFooter={false}
              topOffset="40vmin"
              width={400}
              containerProps={{
                className: 'unblock-dialog',
              }}
            >
              <Pane width="95%" margin="auto">
                <RadioGroup
                  size={16}
                  value={this.state.unblockDialog.selected}
                  options={this.state.unblockDialog.options}
                  onChange={(event) =>
                    this.updateUnblockDialogState({
                      selected: event.target.value,
                    })
                  }
                />
                {this.state.unblockDialog.requirePassword ? (
                  <PasswordPrompt
                    hasHeader={false}
                    hasFooter={false}
                    minWidth="auto"
                    minHeight={50}
                    inputWidth="100%"
                    maxInputWidth="auto"
                    onSuccess={this.unblock}
                  />
                ) : (
                  <Pane
                    display="flex"
                    alignItems="center"
                    justifyContent="center"
                    marginTop={20}
                  >
                    <Button
                      height={32}
                      appearance="primary"
                      iconBefore={UnlockIcon}
                      onClick={this.unblock}
                    >
                      {translate('unblock')}
                    </Button>
                  </Pane>
                )}
              </Pane>
            </Dialog>
          </Fragment>
        )}
      </Fragment>
    );
  }
}
