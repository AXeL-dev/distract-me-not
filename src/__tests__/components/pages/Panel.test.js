import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Panel } from 'components';

// Mock Chrome storage with default values
const mockStorageData = {
  isEnabled: true,
  mode: 'denylist',
  denyList: [],
  allowList: [],
  unblock: {
    isEnabled: false,
  },
  logs: {
    isEnabled: false,
  },
  timer: {
    isEnabled: false,
  },
  schedule: {
    isEnabled: false,
    days: {},
  },
  hideReportIssueButton: false,
  showAddWebsitePrompt: false,
};

beforeEach(() => {
  // Reset Chrome storage mocks
  global.chrome.storage.local.get.mockResolvedValue(mockStorageData);
  global.chrome.storage.sync.get.mockResolvedValue({});
});

it('renders panel correctly', () => {
  const { container, asFragment } = render(<Panel />);
  const status = screen.getByText(/status/i);
  const mode = screen.getByText(/mode/i);
  expect(container).not.toBeEmptyDOMElement();
  expect(status).toBeInTheDocument();
  expect(mode).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

it('updates status when toggled', () => {
  render(<Panel />);
  const statusSwitch = screen.getByRole('checkbox');
  const oldStatus = statusSwitch.checked;
  fireEvent.click(statusSwitch);
  expect(statusSwitch.checked).toBe(!oldStatus);
});

it('updates mode when changed to allow list', () => {
  render(<Panel />);
  const allowListMode = screen.getByText(/allowList/i);
  fireEvent.click(allowListMode);
  const allowListButton = screen.getByRole('button', {
    name: 'allowList',
  });
  expect(allowListButton.getAttribute('data-checked')).toEqual('true');
});

it('updates mode when changed to deny list', () => {
  render(<Panel />);
  const denyListMode = screen.getByText(/denyList/i);
  fireEvent.click(denyListMode);
  const denyListButton = screen.getByRole('button', {
    name: 'denyList',
  });
  expect(denyListButton.getAttribute('data-checked')).toEqual('true');
});
