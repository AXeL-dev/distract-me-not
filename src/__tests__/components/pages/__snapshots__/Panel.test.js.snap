// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders panel correctly 1`] = `
<DocumentFragment>
  <div
    class="ub-min-w_350px ub-box-szg_border-box"
  >
    <div
      class="ub-b-btm_1px-solid-E6E8F0 ub-dspl_flex ub-algn-itms_center ub-just-cnt_center ub-h_62px ub-box-szg_border-box"
    >
      <img
        alt="logo"
        class="logo"
        src="icons/magnet-256.png"
      />
      <h2
        class="cursor-default ub-mt_0px ub-mb_0px ub-color_333 ub-fnt-fam_v92vt9 ub-f-wght_bold ub-fnt-sze_18px ub-ln-ht_24px ub-ltr-spc_-0-07px ub-box-szg_border-box"
      >
        appName
      </h2>
    </div>
    <div
      class="ub-dspl_flex ub-flx-drct_column ub-min-h_115px ub-pl_16px ub-pr_16px ub-pb_20px ub-pt_20px ub-box-szg_border-box"
    >
      <div
        class="ub-dspl_flex ub-box-szg_border-box"
      >
        <div
          class="ub-dspl_flex ub-algn-itms_center ub-flx_1 ub-box-szg_border-box"
        >
          <span
            class="cursor-default ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-fnt-fam_b77syt ub-color_474d66 ub-box-szg_border-box"
          >
            status
          </span>
        </div>
        <div
          class="ub-dspl_flex ub-algn-itms_center ub-mr_0px ub-box-szg_border-box"
        >
          <label
            class="ub-dspl_block ub-w_36px ub-pst_relative ub-box-szg_border-box"
          >
            <input
              class="css-1yog12j ub-b-btm_0 ub-b-lft_0 ub-b-rgt_0 ub-b-top_0 ub-h_1px ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pb_0 ub-pl_0 ub-pr_0 ub-pt_0 ub-pst_absolute ub-wht-spc_nowrap ub-w_1px ub-opct_0 ub-box-szg_border-box"
              type="checkbox"
            />
            <div
              class="ub-h_18px ub-w_36px ub-bblr_9999px ub-bbrr_9999px ub-btlr_9999px ub-btrr_9999px ub-crsr_pointer ub-box-szg_border-box"
            >
              <div
                class="css-17bhmw6 ub-h_18px ub-w_18px ub-box-szg_border-box"
                data-checked="false"
              />
              <div
                class="css-1rprxtl ub-w_36px ub-dspl_flex ub-box-szg_border-box"
                data-checked="false"
              >
                <div
                  class="ub-flx_1 ub-pb_2px ub-pl_2px ub-pr_2px ub-pt_2px ub-box-szg_border-box"
                >
                  <div
                    class="css-13qcn8a ub-w_14px ub-h_14px ub-box-szg_border-box"
                  />
                </div>
              </div>
            </div>
          </label>
        </div>
      </div>
      <div
        class="ub-dspl_flex ub-pt_20px ub-gap_10px ub-box-szg_border-box"
      >
        <div
          class="ub-dspl_flex ub-algn-itms_center ub-flx_1 ub-box-szg_border-box"
        >
          <span
            class="cursor-default ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-fnt-fam_b77syt ub-color_474d66 ub-box-szg_border-box"
          >
            mode
          </span>
        </div>
        <div
          class="ub-dspl_flex ub-algn-itms_center ub-min-w_0px ub-box-szg_border-box"
        >
          <div
            class="css-l6zppx ub-dspl_inline-flex ub-max-w_260px ub-box-szg_border-box"
            role="group"
          >
            <button
              class="css-12b25td ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-c1c4d6 ub-b-lft_1px-solid-c1c4d6 ub-b-rgt_1px-solid-c1c4d6 ub-b-top_1px-solid-c1c4d6 ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_32px ub-pl_16px ub-pr_16px ub-bg-clr_white ub-box-szg_border-box"
              data-checked="false"
            >
              <span
                class="overflow-ellipsis"
              >
                blacklist
              </span>
            </button>
            <button
              class="css-12b25td ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-c1c4d6 ub-b-lft_1px-solid-c1c4d6 ub-b-rgt_1px-solid-c1c4d6 ub-b-top_1px-solid-c1c4d6 ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_32px ub-pl_16px ub-pr_16px ub-bg-clr_white ub-box-szg_border-box"
              data-checked="false"
            >
              <span
                class="overflow-ellipsis"
              >
                whitelist
              </span>
            </button>
            <button
              class="css-elp417 ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_white ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_32px ub-pl_16px ub-pr_16px ub-bg-clr_3366FF ub-b-btm-clr_3366FF ub-b-lft-clr_3366FF ub-b-rgt-clr_3366FF ub-b-top-clr_3366FF ub-box-szg_border-box"
              data-checked="true"
            >
              <span
                class="overflow-ellipsis"
              >
                combined
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ub-b-top_1px-solid-E6E8F0 ub-dspl_flex ub-pl_16px ub-pr_16px ub-pb_10px ub-pt_10px ub-algn-itms_center ub-just-cnt_space-between ub-box-szg_border-box"
    >
      <div
        class="ub-dspl_flex ub-gap_10px ub-box-szg_border-box"
      >
        <button
          aria-describedby="evergreen-tooltip-1"
          class="css-1tyjqqv icon-button  ub-pst_relative ub-f-wght_400 ub-dspl_flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_3px ub-bbrr_3px ub-btlr_3px ub-btrr_3px ub-color_474d66 ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_0 ub-pr_0 ub-bg-clr_transparent ub-pb_0 ub-pt_0 ub-w_32px ub-ltr-spc_0 ub-box-szg_border-box"
        >
          <div
            class="ub-dspl_inline-flex ub-box-szg_border-box"
          >
            <svg
              class="css-1wyorxj ub-w_18px ub-h_18px ub-box-szg_border-box"
              data-icon="cog"
              viewBox="0 0 16 16"
            >
              <path
                d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
        <button
          aria-describedby="evergreen-tooltip-2"
          class="css-1tyjqqv icon-button  ub-pst_relative ub-f-wght_400 ub-dspl_flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_3px ub-bbrr_3px ub-btlr_3px ub-btrr_3px ub-color_474d66 ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_0 ub-pr_0 ub-bg-clr_transparent ub-pb_0 ub-pt_0 ub-w_32px ub-ltr-spc_0 ub-box-szg_border-box"
        >
          <div
            class="ub-dspl_inline-flex ub-box-szg_border-box"
          >
            <svg
              class="css-1wyorxj ub-w_18px ub-h_18px ub-box-szg_border-box"
              data-icon="issue-new"
              viewBox="0 0 16 16"
            >
              <path
                d="M10.568.421c-.01.04-.018.08-.026.121-.837.156-1.53.73-1.85 1.497a6 6 0 105.27 5.273 2.51 2.51 0 001.496-1.854c.04-.008.081-.016.121-.026A8 8 0 1110.568.421zM9 12H7v-2h2v2zm0-3H7V4h2v5zm1-6c0-.55.45-1 1-1h1V1c0-.55.45-1 1-1s1 .45 1 1v1h1c.55 0 1 .45 1 1s-.45 1-1 1h-1v1.005c0 .55-.45 1-1 1s-1-.45-1-1V4h-1c-.55 0-1-.45-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        class="ub-box-szg_border-box"
      />
    </div>
  </div>
</DocumentFragment>
`;
